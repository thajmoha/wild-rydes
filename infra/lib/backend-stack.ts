import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as path from "path";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as iam from "aws-cdk-lib/aws-iam";
import * as apigateway from "aws-cdk-lib/aws-apigateway";
import * as cognito from "aws-cdk-lib/aws-cognito";

interface BackendProps extends cdk.StackProps {
  appName: string;
}

export class BackendStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: BackendProps) {
    super(scope, id, props);

    // dynamodb
    const dynamoDbTable = new dynamodb.TableV2(
      this,
      `${props.appName}RidesTable`,
      {
        tableName: `${props.appName}RidesTable`,
        partitionKey: { name: "RideId", type: dynamodb.AttributeType.STRING },
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      }
    );

    // lambda
    const myLambda = new lambda.Function(
      this,
      `${props.appName}RequestUnicorn`,
      {
        functionName: `${props.appName}RequestUnicorn`,
        runtime: lambda.Runtime.NODEJS_16_X,
        code: lambda.Code.fromAsset(
          path.join(__dirname, "lambda-backend-handler")
        ),
        handler: "requestUnicorn.handler",
      }
    );

    // add lambda permission
    const dynamodbPolicy = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["dynamodb:PutItem"],
      resources: [dynamoDbTable.tableArn],
    });
    myLambda.addToRolePolicy(dynamodbPolicy);
    myLambda.role?.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName("AmazonDynamoDBFullAccess")
    );
    myLambda.role?.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName(
        "service-role/AWSLambdaBasicExecutionRole"
      )
    );

    // apigateway
    const api = new apigateway.RestApi(this, `${props.appName}Api`, {
      restApiName: `${props.appName}Api`,
      endpointTypes: [apigateway.EndpointType.EDGE],
    });
    // authorization
    const userPoolId = cdk.Fn.importValue("UserPoolId");

    const auth = new apigateway.CognitoUserPoolsAuthorizer(
      this,
      `${props.appName}ApiAuth`,
      {
        cognitoUserPools: [
          cognito.UserPool.fromUserPoolId(this, "ImportedUserPool", userPoolId),
        ],
      }
    );
    // ride resource
    const rideResource = api.root.addResource("ride");
    // Enable CORS for the ride resource
    rideResource.addCorsPreflight({
      allowOrigins: apigateway.Cors.ALL_ORIGINS,
      allowMethods: apigateway.Cors.ALL_METHODS,
    });
    rideResource.addMethod("POST", new apigateway.LambdaIntegration(myLambda), {
      authorizer: auth,
      authorizationType: apigateway.AuthorizationType.COGNITO,
    });

    new cdk.CfnOutput(this, "prodApiUrl", {
      value: api.url,
      exportName: "prodApiUrl",
    });
  }
}
