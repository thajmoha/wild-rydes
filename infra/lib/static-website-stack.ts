import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as iam from "aws-cdk-lib/aws-iam";
import * as s3deploy from "aws-cdk-lib/aws-s3-deployment";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as path from "path";
import * as eventsources from "aws-cdk-lib/aws-lambda-event-sources";

interface StaticWebsiteProps extends cdk.StackProps {
  rootDomainName: string;
  appName: string;
}

export class StaticWebsiteStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: StaticWebsiteProps) {
    super(scope, id, props);

    const logsBucket = new s3.Bucket(this, "logs", {
      bucketName: `logs.${props.rootDomainName}`,
      autoDeleteObjects: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // define bucket metrics
    const websiteBucketMetrics: s3.BucketMetrics[] = [{ id: "UsageFilter" }];

    const websiteBucket = new s3.Bucket(this, "website", {
      bucketName: props.rootDomainName,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      metrics: websiteBucketMetrics,
      websiteErrorDocument: "error.html",
      websiteIndexDocument: "index.html",
      serverAccessLogsBucket: logsBucket,
      autoDeleteObjects: true,
      blockPublicAccess: {
        blockPublicAcls: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
        blockPublicPolicy: false,
      },
    });
    websiteBucket.addToResourcePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["s3:GetObject"],
        resources: [websiteBucket.arnForObjects("*")],
        principals: [new iam.AnyPrincipal()],
      })
    );

    new s3deploy.BucketDeployment(this, "DeployWebsite", {
      sources: [s3deploy.Source.asset("../Wild_Rydes")],
      destinationBucket: websiteBucket,
    });
    console.log("----------------------");
    const userPoolId = cdk.Fn.importValue("UserPoolId");
    const userPoolClientId = cdk.Fn.importValue("UserPoolClientId");
    const apiInvokeUrl = cdk.Fn.importValue("prodApiUrl");
    console.log("----------------------");

    const myLambda = new lambda.Function(this, "lambda-config-updates", {
      functionName: "WildRydesConfigUpdatedFunction",
      runtime: lambda.Runtime.PYTHON_3_10,
      code: lambda.Code.fromAsset(
        path.join(__dirname, "lambda-config-update-handler")
      ),
      handler: "main.lambda_handler",
      environment: {
        appName: props.appName,
        userPoolId: userPoolId,
        userPoolClientId: userPoolClientId,
        invokeUrl: apiInvokeUrl,
      },
    });

    // Custom IAM policy
    const s3Policy = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["s3:GetObject", "s3:PutObject"],
      resources: ["*"], // Change this to your S3 bucket ARN
    });
    const cognitoPolicy = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["cognito-idp:ListUserPools", "cognito-idp:ListUserPoolClients"],
      resources: ["*"],
    });
    myLambda.addToRolePolicy(s3Policy);
    myLambda.addToRolePolicy(cognitoPolicy);

    myLambda.addEventSource(
      new eventsources.S3EventSource(websiteBucket, {
        events: [s3.EventType.OBJECT_CREATED],
        filters: [{ prefix: "js/config_template.js" }],
      })
    );
  }
}
