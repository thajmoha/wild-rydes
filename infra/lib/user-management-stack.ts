//see fore more info
// https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito-readme.html#user-pools

import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as cognito from "aws-cdk-lib/aws-cognito";
import { Duration } from "aws-cdk-lib";

interface UserManagementProps extends cdk.StackProps {
  appName: string;
}

export class UserManagementStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: UserManagementProps) {
    super(scope, id, props);

    const userPoolName = `${props.appName}UserPool`;
    const pool = new cognito.UserPool(this, `${props.appName}Pool`, {
      userPoolName: userPoolName,
      deletionProtection: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      // [sign up]
      // Users can either be signed up by the app's administrators or can sign themselves up
      // for this app user can sign themselves up

      selfSignUpEnabled: true,
      // see userVerification for signup configuration
      autoVerify: {
        email: true,
      },
      keepOriginal: {
        email: true,
      },
      standardAttributes: {
        email: {
          required: true,
          mutable: false,
        },
      },

      // [sign in]
      // Users registering or signing in into your application can do so with multiple identifiers.
      // There are 4 options available: username, email, phone, preferredUsername
      signInAliases: {
        username: true,
      },
      signInCaseSensitive: false, // case insensitive is preferred in most situations

      // [message delivery]
      // By default, user pools are configured to use Cognito's built in email capability, which will send emails from no-reply@verificationemail.com
      // email: cognito.UserPoolEmail.withCognito('support@myawesomeapp.com'),

      // [security]
      accountRecovery: cognito.AccountRecovery.EMAIL_ONLY,
      // the following parameters are the default values - explicit here for information
      passwordPolicy: {
        minLength: 8,
        requireLowercase: true,
        requireUppercase: true,
        requireDigits: true,
        requireSymbols: true,
        tempPasswordValidity: Duration.days(7),
      },
      mfa: cognito.Mfa.OFF,
    });
    pool.addClient(`${userPoolName}-app-client`, {
      userPoolClientName: `${props.appName}WebApp`,
    });
  }
}
