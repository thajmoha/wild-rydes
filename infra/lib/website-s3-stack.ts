import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as iam from "aws-cdk-lib/aws-iam";
import * as s3deploy from "aws-cdk-lib/aws-s3-deployment";

interface WebsiteS3Props extends cdk.StackProps {
  rootDomainName: string;
}

export class WebsiteS3Stack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: WebsiteS3Props) {
    super(scope, id, props);

    const logsBuckt = new s3.Bucket(this, "logs", {
      bucketName: `logs.${props.rootDomainName}`,
      autoDeleteObjects: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // define bucket metrics
    const websiteBucketMetrics: s3.BucketMetrics[] = [{ id: "UsageFilter" }];

    const websiteBucket = new s3.Bucket(this, "website", {
      bucketName: props.rootDomainName,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      metrics: websiteBucketMetrics,
      websiteErrorDocument: "error.html",
      websiteIndexDocument: "index.html",
      serverAccessLogsBucket: logsBuckt,
      autoDeleteObjects: true,
      blockPublicAccess: {
        blockPublicAcls: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
        blockPublicPolicy: false,
      },
    });
    websiteBucket.addToResourcePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["s3:GetObject"],
        resources: [websiteBucket.arnForObjects("*")],
        principals: [new iam.AnyPrincipal()],
      })
    );

    new s3deploy.BucketDeployment(this, "DeployWebsite", {
      sources: [s3deploy.Source.asset("../Wild_Rydes")],
      destinationBucket: websiteBucket,
    });
  }
}
