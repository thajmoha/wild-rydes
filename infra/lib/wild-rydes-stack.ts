import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as iam from "aws-cdk-lib/aws-iam";
import * as s3deploy from "aws-cdk-lib/aws-s3-deployment";

interface WildRydesStackProps extends cdk.StackProps {
  rootDomainName: string;
}

export class WildRydesStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: WildRydesStackProps) {
    super(scope, id, props);

    const logsBuckt = new s3.Bucket(this, "logs", {
      bucketName: `logs.${props.rootDomainName}`,
      autoDeleteObjects: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const websiteBucket = new s3.Bucket(this, "website", {
      bucketName: props.rootDomainName,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      websiteErrorDocument: "error.html",
      websiteIndexDocument: "index.html",
      serverAccessLogsBucket: logsBuckt,
      autoDeleteObjects: true,
      blockPublicAccess: {
        blockPublicAcls: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
        blockPublicPolicy: false,
      },
    });
    websiteBucket.addToResourcePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["s3:GetObject"],
        resources: [websiteBucket.arnForObjects("*")],
        principals: [new iam.AnyPrincipal()],
      })
    );

    new s3deploy.BucketDeployment(this, "DeployWebsite", {
      sources: [s3deploy.Source.asset("../Wild_Rydes")],
      destinationBucket: websiteBucket,
    });
  }
}
