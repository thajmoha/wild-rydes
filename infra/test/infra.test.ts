import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { WildRydesStack } from "../lib/wild-rydes-stack";

test("stack created", () => {
  const app = new cdk.App();
  const stack = new WildRydesStack(app, "MyTestStack", {
    env: { account: "test-account", region: "us-east-1" },
    rootDomainName: "example",
  });
  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
