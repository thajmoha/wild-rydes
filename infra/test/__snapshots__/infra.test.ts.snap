// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stack created 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-test-account-us-east-1",
          "S3Key": "e976a796f036a5efbf44b99e44cfb5a961df08d8dbf7cd37e60bf216fb982a00.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "DeployWebsiteAwsCliLayer17DBC421",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-test-account-us-east-1",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-test-account-us-east-1/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "website333B7473",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "website333B7473",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-test-account-us-east-1",
          "S3Key": "b7f33614a69548d6bafe224d751a7ef238cde19097415e553fe8b63a4c8fd8a6.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "logs0B6081B1",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DeployWebsiteAwsCliLayer17DBC421": {
      "Properties": {
        "Content": {
          "S3Bucket": "cdk-hnb659fds-assets-test-account-us-east-1",
          "S3Key": "3322b7049fb0ed2b7cbb644a2ada8d1116ff80c32dca89e6ada846b5de26f961.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "DeployWebsiteCustomResourceD116527B": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Ref": "website333B7473",
        },
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          "cdk-hnb659fds-assets-test-account-us-east-1",
        ],
        "SourceObjectKeys": [
          "c277ca827943ac0f48ee6e57140a208e6e22d6d93e57ba383b0740023612330f.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "logs0B6081B1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccessControl": "LogDeliveryWrite",
        "BucketName": "logs.example",
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "logsAutoDeleteObjectsCustomResourceA82AC6E1": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "logsPolicy42E8B048",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "logs0B6081B1",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "logsPolicy42E8B048": {
      "Properties": {
        "Bucket": {
          "Ref": "logs0B6081B1",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "logs0B6081B1",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "logs0B6081B1",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "website333B7473": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketName": "example",
        "LoggingConfiguration": {
          "DestinationBucketName": {
            "Ref": "logs0B6081B1",
          },
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": false,
          "BlockPublicPolicy": false,
          "IgnorePublicAcls": false,
          "RestrictPublicBuckets": false,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
          {
            "Key": "aws-cdk:cr-owned:0bbd7f95",
            "Value": "true",
          },
        ],
        "WebsiteConfiguration": {
          "ErrorDocument": "error.html",
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "websiteAutoDeleteObjectsCustomResource15116AC6": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "websitePolicy3ACEADD2",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "website333B7473",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "websitePolicy3ACEADD2": {
      "Properties": {
        "Bucket": {
          "Ref": "website333B7473",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "website333B7473",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "website333B7473",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "website333B7473",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
